src/time_integration/paired_explicit_runge_kutta/methods_PERK3.jl

```diff
@@ -0,0 +1,409 @@
+Multirate Time-Integration based on Dynamic ODE Partitioning through Adaptively Refined Meshes for Compressible Fluid Dynamics
+[Arxiv: 10.48550/arXiv.2403.05144](https://doi.org/10.48550/arXiv.2403.05144)
+"""
+mutable struct PairedExplicitRK3 <: AbstractPairedExplicitRKSingle
+    const num_stages::Int
+
+    a_matrix::Matrix{Float64}
+    c::Vector{Float64}
+    dt_opt::Float64
+end # struct PairedExplicitRK3
+
+# Constructor for previously computed A Coeffs
+function PairedExplicitRK3(num_stages, base_path_a_coeffs::AbstractString,
+                           dt_opt;
+                           cS2 = 1.0)
+    a_matrix, c = compute_PairedExplicitRK3_butcher_tableau(num_stages,
+                                                            base_path_a_coeffs;
+                                                            cS2)
+
+    return PairedExplicitRK3(num_stages, a_matrix, c, dt_opt)
+end
+
+# Constructor that computes Butcher matrix A coefficients from a semidiscretization
+function PairedExplicitRK3(num_stages, tspan, semi::AbstractSemidiscretization;
+                           verbose = false, cS2 = 1.0)
+    eig_vals = eigvals(jacobian_ad_forward(semi))
+
+    return PairedExplicitRK3(num_stages, tspan, eig_vals; verbose, cS2)
+end
+
+# Constructor that calculates the coefficients with polynomial optimizer from a list of eigenvalues
+function PairedExplicitRK3(num_stages, tspan, eig_vals::Vector{ComplexF64};
+                           verbose = false, cS2 = 1.0)
+    a_matrix, c, dt_opt = compute_PairedExplicitRK3_butcher_tableau(num_stages,
+                                                                    tspan,
+                                                                    eig_vals;
+                                                                    verbose, cS2)
+    return PairedExplicitRK3(num_stages, a_matrix, c, dt_opt)
+end
+
+# This struct is needed to fake https://github.com/SciML/OrdinaryDiffEq.jl/blob/0c2048a502101647ac35faabd80da8a5645beac7/src/integrators/type.jl#L77
+# This implements the interface components described at
+# https://diffeq.sciml.ai/v6.8/basics/integrator/#Handing-Integrators-1
+# which are used in Trixi.
+mutable struct PairedExplicitRK3Integrator{RealT <: Real, uType, Params, Sol, F, Alg,
+                                           PairedExplicitRKOptions} <:
+               AbstractPairedExplicitRKSingleIntegrator
+    u::uType
+    du::uType
+    u_tmp::uType
+    t::RealT
+    dt::RealT # current time step
+    dtcache::RealT # ignored
+    iter::Int # current number of time steps (iteration)
+    p::Params # will be the semidiscretization from Trixi
+    sol::Sol # faked
+    f::F
+    alg::Alg # This is our own class written above; Abbreviation for ALGorithm
+    opts::PairedExplicitRKOptions
+    finalstep::Bool # added for convenience
+    # PairedExplicitRK stages:
+    k1::uType
+    k_higher::uType
+    k_s1::uType # Required for custom third order version of PairedExplicitRK3
```

Comment:
Let's see what the others say, but maybe we should rename this also.
At least for consistency maybe to `k_S1`.

Reply:
I don't seem to fully grasp what the role of this variable is just from looking at the code. I'll look up the literature for this since there is also a question mark in some lines. But could you quickly clarify what u mean by this?

Line 364

```
# IDEA: Stop for loop at num_stages -1 to avoid if (maybe more performant?)
                if stage == alg.num_stages - 1
                    @threaded for i in eachindex(integrator.du)
                        integrator.k_s1[i] = integrator.k_higher[i]
                    end
                end
```

Reply:
Essentially `k_s1` is the variable for $K_{S-1}$ which is the second-last stage. This also explains the bit of code you see there, which copies at $i = S-1$ the values from `k_higher` (which stores always the newest information) to `k_s1`.

Reply:
Btw., maybe you can try out the TODO? 
I suspect that it won't make a difference as the compiler is hopefully smart enough to figure this out, but you never know unless you try.

Reply:
I will do that and report the result. Thank you for your clarification!

Reply:
Hey, I tried this and it is about 3 times faster, so it's really good! However, another thing also changed with it, which is the result I get from `analysis_callback(sol)` which is now `(l2 = [NaN], linf = [0.0])` even though I changed to `tspan = (0.0, 20.0)`. I'm not so certain why this causes such an improvement. Sounds too good to be true. Is there something wrong?

Reply:
Yes, that sounds like a bug. Three times is way too fast, and as the errors are also blowing up it looks like the functionality changed.

Reply:
So the comment is probably quite misleading: The idea I had there is to stop the loop at $S-1$, then copy in the values to $K_{S-1}$ and then run the last stage. If you omit the last stage completely, the result is gibberish.

Reply:
Oh I see, let me change this to what you just explained.

Reply:
Having done this properly this time, I'd like to report that upon using the benchmark tool, this improves the runtime by about 1-2 seconds.

Reply:
Cool, then let's keep it this way!